influxql/functions.go

```diff
@@ -309,3 +316,322 @@ func (r *IntegerMovingAverageReducer) Emit() []FloatPoint {
+
+	// Dampening parameter
+	phi float64
+
+	// Season period
+	m        int
+	seasonal bool
+
+	// Horizon
+	h int
+
+	// Interval between points
+	interval int64
+
+	// Whether to include all data or only future values
+	includeFitData bool
+
+	// NelderMead optimizer
+	optim *neldermead.Optimizer
+	// Small difference bound for the optimizer
+	epsilon float64
+
+	y      []float64
+	points []FloatPoint
+}
+
+const (
+	defaultAlpha   = 0.5
+	defaultBeta    = 0.5
+	defaultGamma   = 0.5
+	defaultPhi     = 0.5
+	defaultEpsilon = 1.0e-4
+)
+
+// NewFloatHoltWintersReducer creates a new FloatHoltWintersReducer.
+func NewFloatHoltWintersReducer(h, m int, includeFitData bool, interval time.Duration) *FloatHoltWintersReducer {
+	seasonal := true
+	if m < 2 {
+		seasonal = false
+	}
+	return &FloatHoltWintersReducer{
+		alpha:          defaultAlpha,
+		beta:           defaultBeta,
+		gamma:          defaultGamma,
+		phi:            defaultPhi,
+		h:              h,
+		m:              m,
+		seasonal:       seasonal,
+		includeFitData: includeFitData,
+		interval:       int64(interval),
+		optim:          neldermead.New(),
+		epsilon:        defaultEpsilon,
+	}
+}
+
+func (r *FloatHoltWintersReducer) aggregate(time int64, value float64) {
+	r.points = append(r.points, FloatPoint{
+		Time:  time,
+		Value: value,
+	})
+}
+
+// AggregateFloat aggregates a point into the reducer and updates the current window.
+func (r *FloatHoltWintersReducer) AggregateFloat(p *FloatPoint) {
+	r.aggregate(p.Time, p.Value)
+}
+
+// AggregateInteger aggregates a point into the reducer and updates the current window.
+func (r *FloatHoltWintersReducer) AggregateInteger(p *IntegerPoint) {
+	r.aggregate(p.Time, float64(p.Value))
+}
+
+func (r *FloatHoltWintersReducer) roundTime(t int64) int64 {
+	return r.interval * ((t + r.interval/2) / r.interval)
+}
+
+func (r *FloatHoltWintersReducer) Emit() []FloatPoint {
+	if l := len(r.points); l < 2 || r.seasonal && l < r.m {
+		return nil
+	}
+	// First fill in r.y with values and NaNs for missing values
+	start, stop := r.roundTime(r.points[0].Time), r.roundTime(r.points[len(r.points)-1].Time)
+	count := (stop - start) / r.interval
+	r.y = make([]float64, 1, count)
+	r.y[0] = r.points[0].Value
+	t := r.roundTime(r.points[0].Time)
+	for _, p := range r.points[1:] {
+		rounded := r.roundTime(p.Time)
+		if rounded <= t {
+			// Drop values that occur for the same time bucket
+			continue
+		}
+		t += r.interval
+		// Add any missing values before the next point
+		for rounded != t {
+			// Add in a NaN so we can skip it later.
+			r.y = append(r.y, math.NaN())
+			t += r.interval
+		}
+		r.y = append(r.y, p.Value)
+	}
+
+	// Smoothing parameters
+	alpha, beta, gamma := r.alpha, r.beta, r.gamma
+
+	// Seasonality
+	m := r.m
+
+	// Dampening paramter
+	phi := r.phi
+
+	// Starting guesses
+	// NOTE: Since these values are guesses
+	// in the cases where we were missing data,
+	// we can just skip the value and call it good.
+
+	l_0 := 0.0
```

Comment:
I agree but in this case I felt like it added clarity since in many cases the subscript is multiple characters long: `l_tp`, `b_tp`, `s_tm` and `s_tmh` as a few examples.

I feel like `ltp`,`btp`, `stm`, and `stmh` don't quite communicate the correct usage of the vars.

I could be convinced otherwise/am open to better suggestions.